src/evaluator.ml

```diff
@@ -1,31 +1,146 @@
 open Ast
+open Common
 
 
-let rec check_if_dec_in_db (Clause (h, b), db) =
-    match db with
-    | [] -> false
-    | ((Clause (h1, b1)) :: xs) ->
-        if h = h1
-        then true
-        else check_if_dec_in_db (Clause (h,b) ,xs)
+let (fresh, reset) =
+  let nxt = ref 0 in
+  let f () = (nxt := !nxt + 1; string_of_int (!nxt)) in
+  let r () = nxt := 0 in
+  (f, r) 
 
-let rec update_dec_in_db (Clause (h, newb), db) =
-    match db with
-    | [] -> []
-    | ((Clause (h1, b1)) :: xs) ->
-        if h = h1
-        then (Clause (h, newb)) :: xs
-        else (Clause (h1, b1)) :: (update_dec_in_db (Clause (h, newb), xs))
+let found_solution = ref false
+                   
+let rec find_vars q  =
+  match q with [] -> []
+             | (x::xs) -> (match x with VarExp(v) -> [x] @ (find_vars xs)
```

Comment:
You could just use `x :: (find_vars xs)` here.

Reply:
yes I can.